docs/how-to/contribute-code.rst

```diff
@@ -354,18 +458,101 @@ After running the test suite, generate a coverage report by running:
 
       (venv) C:\...>tox -e coverage
 
-You can run the test suite and coverage together by including the testing
-environment to run, e.g. ``py,coverage`` or ``py310,coverage``.
+To run the test suite along with this coverage reporting, run:
+
+.. tabs::
+
+  .. group-tab:: macOS
+
+    .. code-block:: console
+
+      (venv) $ tox p -m test
+
+  .. group-tab:: Linux
+
+    .. code-block:: console
+
+      (venv) $ tox p -m test
+
+  .. group-tab:: Windows
+
+    .. code-block:: doscon
+
+      (venv) C:\...>tox p -m test
+
+Coverage report for host platform
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If all supported versions of Python are available to tox, then coverage for the
```

Comment:
The distinction between `-e coverage` and `-e coverage-platform` isn't clear to me based on this description - if `coverage` fills in the gaps for missing platforms, what is `-platform` for? Also - there's no mention of Python version specific variants. Functionally - how do I get 100% coverage for `tox -e py`?

Reply:
I'm super steeped in this configuration at this point; so, this feedback helps me understand how it's perceived by others. I'll need to take this in to consideration to convey what's happening with some updates.

To answer the question for the time being, though, running `tox -e coverage` leaves both "python version exclusions" as well as "platform exclusions" enabled; running `tox -e coverage-platform` disables "python version exclusions" while keeping "platform exclusions" enabled.

Basically, `tox -e coverage` asks if the current platform and Python version is covered; `tox -e coverage-platform` asks if the current platform is covered for all Python versions.

Reply:
> Basically, `tox -e coverage` asks if the current platform and Python version is covered; `tox -e coverage-platform` asks if the current platform is covered for all Python versions.

So... either I'm misunderstanding your terminology, or... no it doesn't? 

`tox -e py && tox -e coverage` runs a single python version on the current platform... and then fails coverage. `

`tox -e py38,py39,py310,py311 && tox -e coverage` passes coverage when run on a single platform.

What am I missing here?

Reply:
The difference is which conditional coverage rules are in effect when the coverage report is produced.

> `tox -e py && tox -e coverage` runs a single python version on the current platform... and then fails coverage.

This runs with both "platform exclusions" and "Python version exclusions" enabled; so, if the host platform and python version are fully covered, then this succeeds. (However, this has a caveat; `tox -e coverage` will look for the oldest version of Python installed; so, technically, `tox -e py` could run for the version of Python in your virtual environment, but `tox -e coverage` may find an older version of Python that exists on the system. So, safest way to run `tox -e py,coverage` is with an explicit Python version, i.e. `tox -e py311,coverage311`)

> `tox -e py38,py39,py310,py311 && tox -e coverage` passes coverage when run on a single platform.

As above, this also runs with "platform exclusions" and "Python version exclusions" enabled. Producing a coverage report for all Python versions only really makes sense for `tox -e coverage-platform`.

The difference between `tox -e coverage` and `tox -e coverage-platform` is whether "python version exclusions" should be disabled.....as it is for the `-e coverage-platform` command.

To round this out, running `tox -e coverage-project` disables the "platform exclusions" as well as the "python version exclusions; this'll only work, though, if you have coverage files produced on Linux, macOS, and Windows.

Does this help? I'm not sure I can tell where we're missing each other...

Reply:
> Does this help? I'm not sure I can tell where we're missing each other...

Not really... maybe I'm misunderstanding the "double negative" aspect of what "platform exclusions enabled" means. I read it as the `#pragma` lines for no-cover-if-windows (et al) are in effect, so you shouldn't get a coverage miss if "exclusions are enabled".

To take this from another angle: what invocation of coverage would let me run a single test run for Python3.10 on macOS (i.e., `tox -e py310`), and get a 100% coverage report?

And, the inverse - how would I generate a *failing* report out of `tox -e py38,py39,py310,py311` on macOS (i.e., I have coverage for all Python versions, but I haven't run on Linux and Windows)?

Reply:
hmm...I was concerned this would remain as confusing as it was to me when I started this...

First, your understanding of "coverage exclusions" is correct; when an exclusion like `no-cover-if-is-windows` is "enabled", then coverage for that code on Windows is ignored.

Second, though, there are two levels of exclusions:
- Python version exclusions
  - When these are enabled, it doesn't matter which version of Python is used for coverage reporting
  - (however, there is the issue of "phantom branches" with using a different version for reporting coverage versus creating the coverage files.)
- Platform exclusions
  - When these are enabled, it doesn't matter which platform is used for coverage reporting

So, from the user perspective trying to create coverage reports:
- Running `tox -e coverage`
  - This has both "python version" and "platform" exclusions enabled
  - So, it doesn't matter the platform or Python version being used to produce the coverage report
  - (however, again, the caveat that `coverage` should be run using the same version of Python used to produce the coverage files because of how that all works.)
- Running `tox -e coverage-platform`
  - This disables "python version" exclusions...so, for instance, the `no-cover-if-gte-py310` exclusion is disabled (i.e. ignored) and coverage is verified for those code blocks
  - However, "platform" exclusions, like `no-cover-if-is-windows`, remain in effect and that coverage is ignored
- Running `tox -e coverage-project`
  - This disables "python version" _and_ "platform" exclusions; so, all of the conditional coverage rules are ignored

So, specifically:

> what invocation of coverage would let me run a single test run for Python3.10 on macOS (i.e., `tox -e py310`), and get a 100% coverage report?

Running `tox -e coverage310` after `tox -e py310` will produce a 100% coverage because, by default, all coverage exclusions are enabled.

> how would I generate a failing report out of `tox -e py38,py39,py310,py311` on macOS (i.e., I have coverage for all Python versions, but I haven't run on Linux and Windows)?

Running `tox -e coverage-project` would disable _all_ exclusions and detect that linux-specific and windows-specific code was not covered.

What about this? And how much of this should be in the docs? because clearly I've internalized all this but it is far from obvious...

Reply:
> hmm...I was concerned this would remain as confusing as it was to me when I started this...
> 
> First, your understanding of "coverage exclusions" is correct; when an exclusion like `no-cover-if-is-windows` is "enabled", then coverage for that code on Windows is ignored
 
Ok - good to know I haven't missed something fundamental here :-)

> What about this? And how much of this should be in the docs? because clearly I've internalized all this but it is far from obvious...

Eureka - I've finally worked out where the gap is!

`tox -e py` runs the test suite on the current default Python3 (3.11 on my machine). However, `tox -e coverage` runs on the *oldest version of Python installed on your machine* (3.8 for me). 

I was seeing a bunch of coverage failures, and assuming they were all platform/version mismatches, but didn't actually verify that they were caused by Python version/platform branches. It turns out they were all 3.8/3.11 branching discrepancies. Once I worked that out, everything fell into place.

That means on my machine, `tox -e py && tox -e coverage311` and `tox -e py38 && tox -e coverage` both pass clean. `tox -e py311 && tox -e coverage311-platform` fails on the python version branches.

So - the issue is with explaining the differences in "default python version" that is used between `py` and `coverage` targets. The fact that `tox -e py && tox -e coverage` doesn't pass clean (unless you only have a single Python version installed) definitely isn't intuitively obvious (as my ... journey... has revealed :-).

I guess the alternative would be a "test and coverage for a single python version" target, but I'm not sure I see an easy way to add that without another set of complications to the tox config.

Reply:
So, considering all this and your other comment above....I'm thinking we just remove talking about `tox -e coverage` and instead keep it simple for a beginner and just say "use `tox -e py310,coverage310` to produce a coverage report" and mention that "310" should be swapped out for the version of Python they are using.

> I guess the alternative would be a "test and coverage for a single python version" target, but I'm not sure I see an easy way to add that without another set of complications to the tox config.

It wouldn't be too difficult to add additional labels for each Python version; so `tox -m test39`, `tox -m test310`, etc. to run the tests and produce the coverage report. And then, we could just talk about these python version-specific labels in the coverage section.....as well as producing platform and project coverage with `tox -m test-platform` and `tox -m test-project`, respectively.

Ultimately, there's a lot of options in this version of the `tox.ini`....but we don't have to expose it all to anyone up front. An adventurous dev can poke around in `tox.ini` to discover this.

What do you think?

P.S. a side concern is people not noticing the difference between `tox -e` and `tox -m`....especially since tox will run `tox -e test` and kinda look like it's something...but really do nothing at all... I haven't found a way to make tox error for an "unknown/undefined" environment.

Reply:
> So, considering all this and your other comment above....I'm thinking we just remove talking about `tox -e coverage` and instead keep it simple for a beginner and just say "use `tox -e py310,coverage310` to produce a coverage report" and mention that "310" should be swapped out for the version of Python they are using.

Sounds like a good approach to me.

> > I guess the alternative would be a "test and coverage for a single python version" target, but I'm not sure I see an easy way to add that without another set of complications to the tox config.
> 
> It wouldn't be too difficult to add additional labels for each Python version; so `tox -m test39`, `tox -m test310`, etc. to run the tests and produce the coverage report. And then, we could just talk about these python version-specific labels in the coverage section.....as well as producing platform and project coverage with `tox -m test-platform` and `tox -m test-project`, respectively.

This also seems like a good idea; it provides a simple path in to the most likely "just run the tests and give me coverage" use case.

> Ultimately, there's a lot of options in this version of the `tox.ini`....but we don't have to expose it all to anyone up front. An adventurous dev can poke around in `tox.ini` to discover this.
> 
> What do you think?

Definitely agreed. We don't need to document everything in the contributor guide. We just need to document enough to explain the common developer use cases. To my mind, that means:

1. Run the tests quickly to prove they all pass
2. Run the tests on one platform and confirm I've got coverage
3. Run as many versions of the tests as I can on one platform, and give me coverage
4. Do as much of a CI-alike run as I can on one platform.

> P.S. a side concern is people not noticing the difference between `tox -e` and `tox -m`....especially since tox will run `tox -e test` and kinda look like it's something...but really do nothing at all... I haven't found a way to make tox error for an "unknown/undefined" environment.

Yeah - that's not great - but if the "public testing interface" is all `tox -m` calls, then it's a lot less likely that people will get caught out.